src/main/java/org/opensearch/security/filter/SecurityFilter.java

```diff
@@ -384,98 +398,114 @@ private <Request extends ActionRequest, Response extends ActionResponse> void ap
-                                    createIndexRequest.index(),
-                                    alias2Name(createIndexRequest.aliases()),
-                                    request.getClass().getSimpleName(),
-                                    e
-                                );
-                                chain.proceed(task, action, request, listener);
-                            } else {
-                                log.error(
-                                    "Request to create index {} with aliases {} failed, failing {}",
-                                    createIndexRequest.index(),
-                                    alias2Name(createIndexRequest.aliases()),
-                                    request.getClass().getSimpleName(),
-                                    e
-                                );
-                                listener.onFailure(e);
-                            }
-                        }
-                    });
-                }
-            } else {
+            User finalUser = user;
+            Consumer<PrivilegesEvaluatorResponse> handleUnauthorized = response -> {
                 auditLog.logMissingPrivileges(action, request, task);
                 String err;
-                if (!pres.getMissingSecurityRoles().isEmpty()) {
-                    err = String.format("No mapping for %s on roles %s", user, pres.getMissingSecurityRoles());
+                if (!response.getMissingSecurityRoles().isEmpty()) {
+                    err = String.format("No mapping for %s on roles %s", finalUser, response.getMissingSecurityRoles());
                 } else {
                     err = (injectedRoles != null)
                         ? String.format(
                             "no permissions for %s and associated roles %s",
-                            pres.getMissingPrivileges(),
+                            response.getMissingPrivileges(),
                             context.getMappedRoles()
                         )
-                        : String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
+                        : String.format("no permissions for %s and %s", response.getMissingPrivileges(), finalUser);
                 }
                 log.debug(err);
-
                 listener.onFailure(new OpenSearchSecurityException(err, RestStatus.FORBIDDEN));
-            }
+            };
+            // NOTE: Since resource-access evaluation requires fetching documents from index, we make the call async otherwise it would
+            // require blocking transport threads leading to thread exhaustion and request timeouts
+            // We perform the rest of the evaluation as normal if the request is not for resource-access or if the feature is disabled
+            resourceAccessEvaluator.evaluateAsync(request, action, context, ActionListener.wrap(response -> {
```

Comment:
I'd like to see if we can move this down into PrivilegesEvaluator along with the rest of the authorizers like system index and snapshot restore.

I know that one of the problems is `PrivilegesEvaluatorResponse pres = eval.evaluate(context);` expects a return value of `PrivilegesEvaluatorResponse` and not designed to be async.

I'm wondering if it makes sense to change this to either return an `Optional` or just handle the case where it can return `null` and then simply `return`. i.e. if the return value indicates that privilege evaluation is not done, then it means its being done async so stop here and the decision to proceed will be handled in the async call where it either proceeds through the chain or returns early with forbidden error if the user does not have access.

Reply:
There is a markPending() method in PrivilegeEvalResponse class. I tried using it but was not successful. I can give it another try. With rest of PrivilegeEval being synchronous there are only 2 paths forward, as i see:
1. Create a delegating handler that handles rest of the evaluation when feature is disabled or request is not of Resource type
2. Utilize markPending() to indicate that access is currently being evaluated and will be processed by the evaluator. However, due to the nature of PrivilegeEval being synchronous i'm not entirely sure if that will be helpful.

Reply:
I tried with markPending(), the end result will still create a blocking operation since in SecurityFilter we will have to wait for response to be complete before proceeding. I saw the same timeouts in test runs for sample plugin. I think the only path forward is to utilize the current async approach. 

To make the code cleaner, i've moved the logic just before the privilege eval call in Security Filter which checks whether resource evaluator should be used before proceeding to actually use it.